tools/heapsnapshot.py

```diff
@@ -0,0 +1,217 @@
+snapshotMeta = {
+    "node_fields": ["type", "name", "id", "self_size", "edge_count", "trace_node_id", "detachedness"],
+    "node_types": [
+        ["hidden", "array", "string", "object", "code", "closure", "regexp", "number", "native", "synthetic", "concatenated string", "sliced string", "symbol", "bigint", "object shape"],
+        "string", "number", "number", "number", "number", "number"
+    ],
+    "edge_fields": ["type", "name_or_index", "to_node"],
+    "edge_types": [
+        ["context", "element", "property", "internal", "hidden", "shortcut", "weak"],
+        "string_or_number", "node"
+    ],
+    "trace_function_info_fields": ["function_id", "name", "script_name", "script_id", "line", "column"],
+    "trace_node_fields": ["id", "function_info_index", "count", "size", "children"],
+    "sample_fields": ["timestamp_us", "last_assigned_id"],
+    "location_fields": ["object_index", "script_id", "line", "column"],
+}
+
+# These indices refer to the index in the snapshot's metadata header
+nodeTypeToMetaIndex = {type: i for i, type in enumerate(snapshotMeta["node_types"][0])}
+edgeTypeToMetaIndex = {type: i for i, type in enumerate(snapshotMeta["edge_types"][0])}
+
+nodeFieldCount = len(snapshotMeta["node_fields"])
+edgeFieldCount = len(snapshotMeta["edge_fields"])
+
+
+def readAddresses(data):
+    # Ordered list of addresses to ensure the registry is the first node, and also so we can process nodes in index order
+    addresses = []
+    addressToNodeIndex = {}
+
+    def addAddress(address):
+        assert address not in addressToNodeIndex, f"Address already exists in the snapshot: '{address}'"
+        addresses.append(address)
+        addressToNodeIndex[address] = len(addresses) - 1
+
+    # The registry is a special case that needs to be either the first or last node to ensure gc "distances" are calculated correctly
+    registryAddress = data["roots"]["registry"]
+    addAddress(registryAddress)
+
+    for address, obj in data["objects"].items():
+        if address == registryAddress:
+            continue
+        addAddress(address)
+
+    return addresses, addressToNodeIndex
+
+
+def convertToSnapshot(data):
+    addresses, addressToNodeIndex = readAddresses(data)
+
+    # Some notable idiosyncrasies with the heap snapshot format:
+    # 1. The snapshot format contains a flat array of nodes and edges. Oddly, edges must reference the "absolute" index of a node's first element after flattening.
+    # 2. A node's outgoing edges are implicitly represented by a contiguous block of edges in the edges array which correspond to the node's position
+    #    in the nodes array and its edge count. So if the first node has 3 edges, the first 3 edges in the edges array are its edges, and so on.
+
+    nodes = []
+    edges = []
+    strings = []
+
+    stringToSnapshotIndex = {}
+
+    def getUniqueId(address):
+        # Addresses are hexadecimal strings
+        return int(address, 16)
```

Comment:
This approach breaks the "reveal in summary" button (and potentially other small features). I suspect it's due to storing the id as an [int within chrome's C++ layer](https://source.chromium.org/chromium/chromium/src/+/main:out/win-Debug/gen/v8/src/inspector/protocol/HeapProfiler.h;l=37-38?q=getHeapObjectId&ss=chromium%2Fchromium%2Fsrc) and it overflows. I can resolve his by incrementing a counter instead to generate ids, but then the diff view would be broken as that relies on stable ids across snapshots. Maybe there's a hash function I can use here? I'm open to ideas.

Reply:
Added a todo for this as well. I'll follow-up with a PR if I find a nice hash function